proposals/NNNN-method-keypaths.md

```diff
@@ -0,0 +1,667 @@
+
+- **Idioms that rely on contextual type**: When the contextual type is either `KeyPath<Root,_>` or `(Root) -> _`, we can
+    omit `Root` from a key path literal. For example:
+
+    ```swift
+    var names: [String]
+
+    names.map(\String.count)  // OK
+    names.map(\.count)        // Also OK: compiler infers String, like magic!
+    ```
+
+    Unbound method references do not have this feature. Using `.someMember` in a context that could accept an unbound
+    method will cause Swift to look for `.someMember` on the corresponding _function_ type — and since function types
+    have no members, this will fail:
+
+    ```swift
+    names.map(String.reversed)  // OK
+    names.map(.reversed)        // error: type '(String) throws -> _' has no member 'reversed'
+    ```
+
+    The magic of the key path is that, given `\.someMember`, Swift searches for `someMember` on the key path’s `Root`
+    instead of on the type `KeyPath` itself.
+
+    There are some idioms where unbound methods could be useful, but are slightly too awkward without this particular
+    contextual type convenience. For example, as we will demonstrate below, method key paths could help mitigate the
+    [oft](https://forums.swift.org/t/pitch-weak-method-storage-modifiers-aka-weak-references/12161)-[discussed](https://forums.swift.org/t/pitch-introduction-of-weak-unowned-closures/6095)
+    problem of weakly bound method references.
+
+
+## Proposed solution
+
+Allow key paths to reference methods, in addition to the properties and subscripts they currently support:
+
+```swift
+struct Foo {
+    func bar() { }
+
+    func baz(zonk: Int, blotz: String) -> Int { … }
+}
+
+\Foo.bar               // KeyPath<Foo, () -> Void>
+\Foo.baz(zonk:blotz:)  // KeyPath<Foo, (Int,String) -> Int>
+```
+
+Note that applying the key path does not _call_ the method; rather, it creates an unapplied method reference — a closure
+waiting for arguments.
+
+This fixes the ❌ inconsistencies above:
+
+```swift
+exampleURL           .deletingLastPathComponent   // Unapplied method…
+exampleURL[keyPath: \.deletingLastPathComponent]  // ✅ …works as a key path!
+
+exampleURL           .encode(to:)                 // Unapplied method with named arguments…
+exampleURL[keyPath: \.encode(to:)]                // ✅ …works as a key path!
+```
+
+Adding this feature opens up possibilties both pleasant and promising.
+
+### Use cases
+
+#### Chains that mix properties and methods
+
+We now have a way of referencing a member chain that includes both properties _and_ methods; the key path
+`\URL.absoluteString.reversed` now compiles.
+
+Why is such a construct useful? Here is an example from [real-world
+code](https://github.com/bustoutsolutions/siesta/blob/3b5b83447a83bb70237537a93ecca18a6fa256b3/Source/Siesta/Pipeline/PipelineProcessing.swift#L164):
+
+```swift
+stagesAndEntries.prefix(upTo: index)
+    .compactMap { $0.cacheEntry?.remove })  // gather removal actions to perform later
```

Comment:
I think the trailing paren is wrong here.

Reply:
Ah, yes, good catch. It was part of the original context of the code, mistakenly copied.